### PR TITLE
feat(HMS-1429): Add GCP to spec

### PR DESCRIPTION
**Pull request by adiabramovitch**
This PR depends on #468, and adds GCP to spec

### CODE REVIEW

**Comment by lzap** [comment]
Please add examples just like in: https://github.com/RHEnVision/provisioning-backend/pull/557

(Perhaps wait until that one is merged and rebase on top of that.)

File: docs/configure-gcp.md [hunk 1]
```diff
@@ -61,11 +59,8 @@ This is a setup in the account in which the service shall deploy the actual inst
 5. Fill in the Service Account e-mail under New principals.
 6. Choose the permissions the provisioning app needs (TBC): 
 Roles:
- - Service Account User
-  
-Permissions:
-  - compute.instances.insert
-
+   - Service Account User
+   - Compute Admin (We are allowing here more permissions than needed)
```
[the remark this answers is not in the document]
Comment (adiabramovitch) [reply docs/configure-gcp.md:63]: yes, I forgot to put this PR on draft, thanks!

**Comment by akhil-jha** [comment]
@adiabramovitch Just tag me when this PR is ready :) I'd want to get started.

File: docs/configure-gcp.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (mshriver) [reply docs/configure-gcp.md:63]: 50% making sure I understand the setup! :)

**Comment by adiabramovitch** [comment]
> @adiabramovitch Just tag me when this PR is ready :) I'd want to get started.

@akhil-jha I think you can start :)

**Comment by akhil-jha** [comment]
Thanks. I think you're missing for GCP. 
```
        "tags": [
          "Reservation"
        ]
```

**Comment by adiabramovitch** [comment]
> Thanks. I think you're missing for GCP.
> 
> ```
>         "tags": [
>           "Reservation"
>         ]
> ```

Right, updated :)

File: cmd/spec/example_reservation.go [hunk 2]
```diff
@@ -150,6 +150,51 @@ var AzureReservationResponsePayloadDoneExample = payloads.AzureReservationRespon
...
+	Zone:               "us-east-4",
+	MachineType:        "e2-micro",
+	Amount:             1,
+	ImageID:            "08a48fed-de87-40ab-a571-f64e30bd0aa8",
+	LaunchTemplateName: "",
+	Name:               "my-instance",
+	GCPOperationName:   "",
+	PowerOff:           false,
+	Instances: []payloads.InstanceResponse{
+		{InstanceID: "2324343212", Detail: models.ReservationInstanceDetail{
```
Comment (lzap) [inline cmd/spec/example_reservation.go:191]: Numbers seem to be a bit longer, `5293983267767892774` in my case, but ok :-)

File: cmd/spec/example_reservation.go [hunk 3]
```diff
@@ -150,6 +150,51 @@ var AzureReservationResponsePayloadDoneExample = payloads.AzureReservationRespon
...
+	ID:                 1305,
+	PubkeyID:           42,
+	SourceID:           "654321",
+	Zone:               "us-east-4",
+	MachineType:        "e2-micro",
+	Amount:             1,
+	Name:               "my-instance",
+	ImageID:            "08a48fed-de87-40ab-a571-f64e30bd0aa8",
+	LaunchTemplateName: "",
+	GCPOperationName:   "",
```
Comment (lzap) [inline cmd/spec/example_reservation.go:174]: Could you provide example for the operation name and maybe also template name? If we ever commit and push this, we will definitely forget to fill that in when the feature finally lands. Also, operation name I don’t remember what this is for, maybe the example can help.

File: cmd/spec/path.yaml [hunk 4]
```diff
@@ -465,6 +465,31 @@ paths:
...
           $ref: '#/components/responses/InternalError'
+  /reservations/gcp:
+    post:
+      operationId: createGCPReservation
+      tags:
+        - Reservation
+      description: >
+        A reservation is a way to activate a job, keeps all data needed for a job to start.
+        A GCP reservation is a reservation created for a GCP job. Image Builder UUID image
+        is required and needs to be shared with the service account.
```
Comment (lzap) [inline cmd/spec/path.yaml:476]: Please explain the name pattern here, we do not drop a word about it.

File: internal/payloads/reservation_payload.go [hunk 5]
```diff
@@ -215,7 +218,10 @@ type GCPReservationRequestPayload struct {
 	SourceID string `json:"source_id" yaml:"source_id"`
 
 	// Optional launch template name global/instanceTemplates/NAME or empty string
-	LaunchTemplateName string `json:"launch_template_name,omitempty"`
+	LaunchTemplateName string `json:"launch_template_name,omitempty" yaml:"launch_template_name"`
+
+	// Optional name pattern of the instance(s).
+	Name string `json:"name_pattern" yaml:"name_pattern"`
```
Comment (lzap) [inline internal/payloads/reservation_payload.go:224]: This should be refactored, creating different JSON names and Go field type names is something we should definitely avoid.

```suggestion
	NamePattern string `json:"name_pattern" yaml:"name_pattern"`
```

It is fine to keep this as `Name` in the response as long as the JSON/YAML has the same name. But for request, I think it is more appropriate to name it as a pattern.

**Review by lzap — CHANGES_REQUESTED** [review]
Hey, I have some comments.

We should talk on tech discussion on the name pattern precedent, is this something we want to apply for aws/azure too?

File: internal/payloads/reservation_payload.go (hunk 5 above)
Comment (adiabramovitch) [reply internal/payloads/reservation_payload.go:224]: So regarding the Name pattern, in a bulkInsert Request, we do not have a name field but a name pattern field which needs to have the ### sign. 
I will change the GCP naming from Name to NamePattern

Comment (lzap) [reply internal/payloads/reservation_payload.go]: Yeah I get that, it is okay to have Name in response but NamePattern in request. It seems logical and correct to me :-)

File: api/openapi.gen.json [hunk 6]
```diff
@@ -107,6 +107,59 @@
...
+          "pubkey_id": 42,
+          "reservation_id": 1305,
+          "source_id": "654321",
+          "zone": "us-east-4"
+        }
+      },
+      "v1.GCPReservationResponsePayloadPendingExample": {
+        "value": {
+          "amount": 1,
+          "gcp_operation_name": "operation-1686646674436-5fdff07e43209-66146b7e-f3f65ec5",
```
Comment (lzap) [inline api/openapi.gen.json:150]: Interesting so it is some sort of unique id of some backend operation...

**Review by lzap — APPROVED** [review]
Great, let’s merge when IQE is ready.

**Review by lzap — APPROVED** [review]
Yeah this will fix the `nil` issue with steps.

**Review by akhil-jha — APPROVED** [review]

**Pull request merged**